### ISSUE TITLE
Alacritty is consistently slower than urxvt in XQuartz (macOS)

### ISSUE THREAD

**Issue by spaghetti2514** [jwilm/alacritty#205]
>Alacritty’s biggest claim is that it’s the fastest terminal emulator available. If there’s a case where it’s not, then it’s either a bug in Alacritty or a misconfigured system

After a fresh install of Alacritty on OS X, I generated a 38MB text file consisting of 10 million lines of "foo" with the command ```printf "%0.sfoo\n" {1..10000000} > 10mil.txt``` and then attempted to compare the speed of Alacritty to the speed of urxvt running in Xquartz by running ```time cat 10mil.txt``` a few times in each terminal at a normal size after priming caches.

**Times taken in urxvt**
2.833s
2.966s
2.555s
**Times taken in alacritty**
5.611s
5.665s
5.553s

>Alacritty’s performance scales well with screen size. Running at larger resolutions will tip the scale further in Alacritty’s favor.

**Times taken in urxvt (full screen)**
3.348s
3.614s
3.639s
**Times taken in alacritty (full screen)**
12.819s
12.641s
13.320s

I normally wouldn't consider this a bug, but I figure you probably want all the comparisons you can get if you're trying to make alacritty the fastest terminal emulator. Also thought it was better to open a new issue rather than commenting on an existing performance issue both for organization and because the root causes might be totally different.

Let me know if there's any additional information you'd like me to provide.

Edit:
```
$ glxinfo | grep OpenGL
OpenGL vendor string: NVIDIA Corporation
OpenGL renderer string: NVIDIA GeForce 320M OpenGL Engine
OpenGL version string: 2.1 NVIDIA-10.2.12 310.90.10.05b20
OpenGL shading language version string: 1.20
OpenGL extensions:
```

**Comment by jwilm** [comment]
Are you using Mesa?

**Comment by jwilm** [comment]
Thanks for the report, we do want to be aware of all these issues.

**Comment by spaghetti2514** [comment]
>Are you using Mesa?

On OS X? I think Xquartz at least packages the library. No idea if it's being used.

Alacritty is just the normal OS X install though. From my understanding of Mesa it shouldn't be involved at all.

**Comment by jwilm** [comment]
Oh, sorry. I saw `urxvt` and made an assumption without reading carefully. Yeah, this doesn't sound good. I have a few things in mind that may have a big impact here. I'll follow up soon.

**Comment by b52** [comment]
Similar results when I run the same tests in full screen (5376x3024) on my Dell XPS 13 running Archlinux with the modesetting driver:

**urxvt**
> cat 10mil.txt  0.00s user 3.06s system 98% cpu 3.093 total

**alacritty**
> cat 10mil.txt  0.00s user 22.32s system 22% cpu 1:37.24 total

That's quite a difference in urxvt's favour.

**Comment by b52** [comment]
[Another comparison to `gnome-terminal`](https://news.ycombinator.com/item?id=13345825) with similar findings.

**Comment by jwilm** [comment]
@b52 are you using the Mesa driver?

**Comment by manishrjain** [comment]
[Another comparison with Terminator](https://github.com/jwilm/alacritty/issues/225) where Alacritty is slower as well.

**Comment by b52** [comment]
@jwilm I assume so that it is used, yes.

**Comment by tmalsburg** [comment]
I tested on Ubuntu 16.10 with `time hexdump file.txt` in full screen. The size of the file was 65MB.  The numbers below are from the second run to warm up the caches:

Alacritty:
```
real    0m51.136s
user    0m14.060s
sys     0m8.764s
```

Urxvt:
```
real	0m17.579s
user	0m10.396s
sys	0m6.604s
```

**Comment by nykma** [comment]
Here goes my `time cat 10mil.txt` results, tested in ArchLinux on MacBookPro 12,1, Dell U2412M(1920*1200), i3wm, `xft:mononoki:size=8`, all fullscreen:

urxvt:
```
real	0m3.063s
user	0m0.000s
sys	0m2.963s
```

alacritty:
```
real    1m38.291s
user    0m0.003s
sys     0m4.930s
```

BTW, how can I check which graphic driver I'm using? I've installed `xf86-video-intel`, wanna check if this is activated or not.

**Comment by jwilm** [comment]
@nykma `glxinfo | grep OpenGL`

**Comment by Senjai** [comment]
I'm just going to chime in here and say on Ubuntu gnome I get marked increases in speed when using Nvidia drivers from edgers. Gnome terminal is a no contest and I get about a second difference compared to uxvrt

**Comment by Senjai** [comment]
@b52 try out the proprietary drivers

**Comment by jwilm** [comment]
I am hoping to fix the Mesa issue so this part of the perf problem will go away. Although perf can currently be improved by installing the proprietary driver, I don't think it's the project's place to be encouraging that.

**Comment by nykma** [comment]
@jwilm Thanks.

```
OpenGL vendor string: Intel Open Source Technology Center
OpenGL renderer string: Mesa DRI Intel(R) Iris 6100 (Broadwell GT3) 
OpenGL core profile version string: 4.5 (Core Profile) Mesa 13.0.3
OpenGL core profile shading language version string: 4.50
OpenGL core profile context flags: (none)
OpenGL core profile profile mask: core profile
OpenGL core profile extensions:
OpenGL version string: 3.0 Mesa 13.0.3
OpenGL shading language version string: 1.30
OpenGL context flags: (none)
OpenGL extensions:
OpenGL ES profile version string: OpenGL ES 3.1 Mesa 13.0.3
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
```

**Comment by b52** [comment]
I was wondering what the progress on this one is.

I think the claim made in the README is kinda bold, given that there are serious performance issues compared to other terminals on some standard platforms.

**Comment by jwilm** [comment]
Hey @b52,

This issue has actually been resolved. There were several duplicates open at one point, and I must have missed this one after the patch.

Did you experience an issue yourself?

**Comment by spaghetti2514** [comment]
@jwilm,
This hasn't been fixed and shouldn't be closed. I just rebuilt alacritty from head and tried and tried the benchmark as outlined in the original issue, and alacritty is the same speed as before.

I know there have been some people who commented on this issue who were actually having mesa issues, but the original issue has nothing to do with that.

**Comment by jwilm** [comment]
@spaghetti2514 oops, my mistake! I looked up at @b52's last comment and it showed the Mesa driver. Going to update the title so I don't make this mistake again.

**Comment by b52** [comment]
@jwilm I also rebuilt the latest HEAD and still have the same performance issues as outlined in this issue.

**Comment by jwilm** [comment]
@b52 the Mesa driver issue should be resolved. If that's not the case for you, could you please open a separate issue with current perf details?

**Comment by Senjai** [comment]
@jwilm what ref was that fixed in? I'm just curious is all to see sht the issue was

**Comment by jwilm** [comment]
@Senjai d40d61a73629aae797ab9ff8b36938ea1237acc2

**Comment by chrisduerr** [comment]
I'd be curious how this benchmark performs on the scrollback PR. It should be faster than urxvt/st now. At least it has been on my machine.

**Comment by spaghetti2514** [comment]
@chrisduerr Performance of the scrollback branch is significantly worse for me.

For the same ```time cat 10mil.txt``` with primed caches

**urxvt**
2.824s
2.826s
2.984s

**alacritty**
10.714s
10.354s
10.483s

**alacritty (fullscreen)**
20.166s
20.745s
20.518s

Using master, the times taken are more in line with the numbers from my original comment, so this is caused by something in the scrollback branch specifically.

**Comment by chrisduerr** [comment]
@spaghetti2514 Are those numbers inside of tmux? That makes a big difference.

**Comment by spaghetti2514** [comment]
@chrisduerr No. Should they be?

**Comment by spaghetti2514** [comment]
Yes. Updated to the latest stable rustc, git pull'd, checked out scrollback, cargo build --release

**Comment by chrisduerr** [comment]
Hmm, for now I'd wait until #1316 gets fixed, even though that shouldn't be the issue, but it might still help. After that it might make sense to look at this issue again.

**Comment by spaghetti2514** [comment]
Alright. Ping me again when it's fixed and I'll recompile and try again.

**Comment by chrisduerr** [comment]
There has been some in-depth testing for the Scrollback release an my Mesa system and it outperformed URxvt in every benchmark.

The #1316 issue has significantly improved some aspects of it, so re-testing the current master probably makes sense.

@spaghetti2514 if you're still around and want to benchmark it again, please let me know what your results are.

Otherwise this issue can probably be closed since performance when scrolling should now be pretty much identical to URxvt on the latest master.

**Comment by spaghetti2514** [comment]
@chrisduerr It's gotten somewhat faster, especially at larger window sizes, but it's still not as fast as urxvt.
Compiled alacritty from master using the latest rustc nightly with cargo build --release

urxvt: **2.8s**
alacritty: **4.4s**

urxvt (full screen): **3.1s**
alacritty (full screen): **5.9s**

The above are all averages of a few runs
